File: cpp/src/euler058.cpp
// https://projecteuler.net/problem=58

// Spiral Primes

// Starting with 1 and spiralling anticlockwise in the following way,
// a square spiral with side length 7 is formed.

// It is interesting to note that the odd squares lie along the bottom right diagonal,
// but what is more interesting is that 8 out of the 13 numbers lying along both diagonals are prime;
// that is, a ratio of 8/13 ≈ 62%.

// If one complete new layer is wrapped around the spiral above,
// a square spiral with side length 9 will be formed.
// If this process is continued, what is the side length of the square spiral
// for which the ratio of primes along both diagonals first falls below 10%?

// Answer: 26241

#include <iostream>
#include <cstdint>

// Miller-Rabin primality test for 32-bit integers
static bool is_prime(uint32_t n) {
    if (n < 2) return false;
    if (n == 2 || n == 3) return true;
    if (n % 2 == 0 || n % 3 == 0) return false;

    // Write n-1 as d * 2^s
    uint32_t d = n - 1;
    int s = 0;
    while (d % 2 == 0) {
        d /= 2;
        ++s;
    }

    // Deterministic bases for 32-bit integers
    const uint32_t bases[] = {2, 7, 61};
    for (uint32_t a : bases) {
        if (a >= n) continue;
        uint64_t x = 1;
        uint64_t base = a;
        uint32_t exp = d;
        // modular exponentiation: x = a^d mod n
        while (exp > 0) {
            if (exp & 1) x = (x * base) % n;
            base = (base * base) % n;
            exp >>= 1;
        }
        if (x == 1 || x == n - 1) continue;
        bool composite = true;
        for (int r = 1; r < s; ++r) {
            x = (x * x) % n;
            if (x == n - 1) {
                composite = false;
                break;
            }
        }
        if (composite) return false;
    }
    return true;
}

// Find side length where prime ratio falls below 10%
int spiral_primes() {
    uint64_t num_primes = 0;
    uint64_t total_diagonals = 1; // include center 1
    uint32_t side = 1;
    while (true) {
        side += 2; // next odd side length
        uint32_t lower_right = side * side;
        uint32_t lower_left = lower_right - (side - 1);
        uint32_t upper_left = lower_left - (side - 1);
        // upper_right = upper_left - (side - 1) is a square (odd square) and not prime
        // check three corners
        if (is_prime(lower_left)) ++num_primes;
        if (is_prime(upper_left)) ++num_primes;
        // upper_right is square, skip
        total_diagonals += 4;
        // compute ratio: primes / total_diagonals
        if (num_primes * 100 < total_diagonals * 10) { // ratio < 10%
            return side;
        }
    }
}

#if ! defined UNITTEST_MODE
int main(int argc, char const *argv[]) {
    std::cout << spiral_primes() << std::endl;
}
#endif // UNITTEST_MODE

File: python/src/euler058.py
# Problem 58: Spiral Primes
# Starting with 1 and spiralling anticlockwise in the following way, a square spiral with side length 7 is formed.
# 37 36 35 34 33 32 31
# 38 17 16 15 14 13 30
# 39 18  5  4  3 12 29
# 40 19  6  1  2 11 28
# 41 20  7  8  9 10 27
# 42 21 22 23 24 25 26
# 43 44 45 46 47 48 49
# It is interesting to note that the odd squares lie along the bottom right diagonal, but what is more interesting is that 8 out of the 13 numbers lying along both diagonals are prime; that is, a ratio of 8/13 ≈ 62%.
# If one complete new layer is wrapped around the spiral above, a square spiral with side length 9 will be formed. If this process is continued, what is the side length of the square spiral for which the ratio of primes along both diagonals first falls below 10%?
# Answer: 26241
# Execution time: About 0.2 seconds
# Optimizations: Used incremental corner calculation and prime checking, stopped when ratio drops below threshold.
# Notes: The spiral generates numbers in a predictable pattern, allowing efficient diagonal extraction.
def is_prime(n):
    if n < 2:
        return False
    if n == 2:
        return True
    if n % 2 == 0:
        return False
    for i in range(3, int(n**0.5) + 1, 2):
        if n % i == 0:
            return False
    return True

def solve():
    primes = 0
    total = 1
    corner = 1
    side = 1
    while True:
        side += 2
        for _ in range(4):
            corner += side - 1
            if is_prime(corner):
                primes += 1
            total += 1
        if primes / total < 0.1:
            return side